bentoml/_internal/runner/runnable.py

```diff
@@ -24,19 +24,101 @@
 RUNNABLE_METHOD_MARK: str = "_bentoml_runnable_method"
 
 
-class Runnable(ABC):
-    SUPPORT_NVIDIA_GPU: bool
-    SUPPORT_CPU_MULTI_THREADING: bool
+class RunnableMeta(ABCMeta, t.Type[t.Any]):
+    supported_resources: Set[str]
+    supports_multi_threading: bool
 
     methods: dict[str, RunnableMethod[t.Any, t.Any, t.Any]] | None = None
 
-    @abstractmethod
-    def __init__(self, **kwargs: t.Any):
-        ...
+    def __new__(
+        cls,
+        name: str,
+        bases: t.Tuple[type, ...],
+        attr_dict: dict[t.Any, t.Any],
+        *,
+        supported_resources: Set[str] | None = None,
+        supports_multi_threading: bool | None = None,
+        **_kwargs: t.Any,
+    ) -> RunnableMeta:
+        res = super().__new__(cls, name, bases, attr_dict)
+
+        if "SUPPORT_NVIDIA_GPU" in attr_dict:
+            if supported_resources is None:
+                if attr_dict["SUPPORT_NVIDIA_GPU"]:
+                    supported_resources = {"nvidia.com/gpu"}
+                else:
+                    supported_resources = set()
+                logger.warning(
+                    f"{name} is using deprecated 'SUPPORT_NVIDIA_GPU'. Please convert to using 'supported_resources':\n"
+                    f"class {name}(Runnable, supported_resources={supported_resources}):\n"
+                    "    ..."
+                )
+            else:
+                logger.warning(
+                    f"Deprecated 'SUPPORT_NVIDIA_GPU' is being ignored in favor of 'supported_resources' for {name}."
+                )
```

Comment:
We probably want to raise a warning or error if the legacy `SUPPORT_NVIDIA_GPU` and `supported_resources`  do not agree like in the multi-threading case.